jetbrains-core/src/software/aws/toolkits/jetbrains/services/federation/AwsConsoleUrlFactory.kt

```diff
@@ -0,0 +1,130 @@
+import org.apache.http.message.BasicNameValuePair
+import software.amazon.awssdk.auth.credentials.AwsCredentials
+import software.amazon.awssdk.auth.credentials.AwsSessionCredentials
+import software.amazon.awssdk.core.client.config.SdkAdvancedClientOption
+import software.amazon.awssdk.regions.Region
+import software.amazon.awssdk.services.sts.StsClient
+import software.amazon.awssdk.services.sts.StsClientBuilder
+import software.aws.toolkits.core.region.AwsRegion
+import software.aws.toolkits.jetbrains.core.AwsClientManager
+import software.aws.toolkits.jetbrains.core.AwsClientManager.Companion.userAgent
+import software.aws.toolkits.jetbrains.core.AwsSdkClient
+import java.time.Duration
+
+class AwsConsoleUrlFactory(
+    private val httpClientBuilder: HttpClientBuilder = HttpClientBuilder.create(),
+    private val stsClientBuilder: StsClientBuilder = StsClient.builder()
+) {
+    fun consoleTld(region: AwsRegion) = when (region.partitionId) {
+        "aws-cn" -> {
+            "amazonaws.cn"
+        }
+        "aws-gov" -> {
+            "amazonaws-us-gov.com"
+        }
+        "aws" -> {
+            "aws.amazon.com"
+        }
+        else -> throw IllegalStateException("Partition '${region.partitionId}' is not supported")
+    }
+
+    fun federationUrl(region: AwsRegion) = "https://signin.${consoleTld(region)}/federation"
+
+    fun consoleUrl(fragment: String? = null, region: AwsRegion): String {
+        val regionSubdomain = when (region.id) {
+            // cn-north-1 is not valid, but is the default...
+            "cn-north-1" -> ""
+            else -> "${region.id}."
+        }
+
+        val consoleHome = "https://${regionSubdomain}console.${consoleTld(region)}"
+
+        return "$consoleHome${fragment ?: "/"}"
+    }
+
+    fun getSigninToken(credentials: AwsCredentials, region: AwsRegion): String {
+        val creds = if (credentials !is AwsSessionCredentials) {
+            val stsClient = stsClientBuilder
+                .httpClient(AwsSdkClient.getInstance().sharedSdkClient())
+                .region(Region.of(region.id))
+                .credentialsProvider { credentials }
+                .overrideConfiguration { configuration ->
+                    userAgent.let { configuration.putAdvancedOption(SdkAdvancedClientOption.USER_AGENT_PREFIX, it) }
+                }
+                .build()
+
+            val tokenResponse = stsClient.use {
+                it.getFederationToken {
+                    it.durationSeconds(Duration.ofMinutes(15).toSeconds().toInt())
+                    it.name("FederatedLoginFromAWSToolkit")
+                }
+            }
+            tokenResponse.credentials().let { AwsSessionCredentials.create(it.accessKeyId(), it.secretAccessKey(), it.sessionToken()) }
+        } else {
+            credentials
+        }
+
+        val sessionJson = """
+            {"sessionId":"${creds.accessKeyId()}","sessionKey":"${creds.secretAccessKey()}","sessionToken": "${creds.sessionToken()}"}
```

Comment:
Can we do this with an object (or even just a map) and Jackson, we've had problems during pen testing with "Json injections" before